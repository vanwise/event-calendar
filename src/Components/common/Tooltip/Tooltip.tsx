import styled, { CSSProp, keyframes } from 'styled-components/macro';
import { AlertIcon } from 'Components/svg';

interface TooltipProps extends WithClassName {
  text: string;
  iconCSS?: CSSProp;
}

function Tooltip({ text, iconCSS, className }: TooltipProps) {
  return (
    <Root className={className}>
      <Icon $CSS={iconCSS} />
      <Text>{text}</Text>
    </Root>
  );
}

const fadeInAnimation = keyframes`
  from {
    opacity: 0;
  }
  to {
    opacity: 1;
    transform: translate(-50%, -10px);
  }
`;

const Text = styled.span`
  position: absolute;
  z-index: 10;
  bottom: calc(100% + 5px);
  left: 50%;
  display: none;
  padding: 10px 15px;
  width: max-content;
  max-width: 300px;
  transform: translateX(-50%);
  border: 1px solid var(--violet);
  border-radius: 20px;
  background: white;
  box-shadow: 0 5px 20px 5px var(--gray-opacity);
  animation: ${fadeInAnimation} 0.3s forwards;
  hyphens: auto;

  &::before {
    content: '';
    position: absolute;
    left: 50%;
    bottom: -19px;
    transform: translateX(-50%);
    border-top: 12px solid var(--violet);
    border-left: 7px solid transparent;
    border-right: 7px solid transparent;
    border-bottom: 7px solid transparent;
  }

  &::after {
    content: '';
    position: absolute;
    left: 50%;
    bottom: -19px;
    transform: translateX(-50%);
    border-top: 10px solid white;
    border-left: 6px solid transparent;
    border-right: 6px solid transparent;
    border-bottom: 9px solid transparent;
  }
`;

const Root = styled.span`
  position: relative;

  &:hover {
    ${Text} {
      display: block;
    }
  }
`;

const Icon = styled(AlertIcon)<{ $CSS?: CSSProp }>`
  ${({ $CSS }) => $CSS}
`;

export default Tooltip;
