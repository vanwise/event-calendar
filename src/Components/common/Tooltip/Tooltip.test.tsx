import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Tooltip from './Tooltip';

const TOOLTIP_MOCK_TEXT = 'tooltip mock text';

function setup(props = {}) {
  return render(<Tooltip text={TOOLTIP_MOCK_TEXT} {...props} />);
}

describe('Tooltip', () => {
  it('should be visible after hover and show mock text', async () => {
    setup();

    const tooltipText = screen.getByText(TOOLTIP_MOCK_TEXT);
    expect(tooltipText).not.toBeVisible();

    const tooltip = screen.getByRole('tooltip', { name: TOOLTIP_MOCK_TEXT });
    await userEvent.hover(tooltip);
    expect(tooltipText).toMatchSnapshot();
  });
});
