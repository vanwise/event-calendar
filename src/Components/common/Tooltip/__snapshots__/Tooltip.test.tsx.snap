// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip should be visible after hover and show mock text 1`] = `
.c1 {
  position: absolute;
  z-index: 10;
  bottom: calc(100% + 5px);
  left: 50%;
  display: none;
  padding: 10px 15px;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  max-width: 300px;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
  border: 1px solid var(--violet);
  border-radius: 20px;
  background: white;
  box-shadow: 0 5px 20px 5px var(--gray-opacity);
  -webkit-animation: fSBaHY 0.3s forwards;
  animation: fSBaHY 0.3s forwards;
  -webkit-hyphens: auto;
  -moz-hyphens: auto;
  -ms-hyphens: auto;
  hyphens: auto;
}

.c1::before {
  content: '';
  position: absolute;
  left: 50%;
  bottom: -19px;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
  border-top: 12px solid var(--violet);
  border-left: 7px solid transparent;
  border-right: 7px solid transparent;
  border-bottom: 7px solid transparent;
}

.c1::after {
  content: '';
  position: absolute;
  left: 50%;
  bottom: -19px;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
  border-top: 10px solid white;
  border-left: 6px solid transparent;
  border-right: 6px solid transparent;
  border-bottom: 9px solid transparent;
}

.c2:hover .c0 {
  display: block;
}

<span
  class="c0 c1"
>
  tooltip mock text
</span>
`;
